src/timestamp.rs

```diff
@@ -0,0 +1,231 @@
+/// Contains the `Timestamp` struct and `ClockSequence` traits
+/// as well as an implementation of ClockSequence for Context (only for features v1 and v6)
+
+/// The number of 100 ns ticks between the UUID epoch
+/// `1582-10-15 00:00:00` and the Unix epoch `1970-01-01 00:00:00`.
+pub const UUID_TICKS_BETWEEN_EPOCHS: u64 = 0x01B2_1DD2_1381_4000;
+/// Stores the number of seconds since epoch,
+/// as well as the fractional nanoseconds of that second
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Timestamp {
+    pub(crate) seconds: u64,
+    pub(crate) nanos: u32,
+    #[cfg(any(feature = "v1", feature = "v6"))]
+    pub(crate) counter: u16,
+}
+
+impl Timestamp {
+    /// Construct a `Timestamp` from its raw component values: an RFC4122
+    /// timestamp and counter.
+    ///
+    /// RFC4122, which defines the V1 UUID, specifies a 60-byte timestamp format
+    /// as the number of 100-nanosecond intervals elapsed since 00:00:00.00,
+    /// 15 Oct 1582, "the date of the Gregorian reform of the Christian
+    /// calendar."
+    pub const fn from_rfc4122(ticks: u64, _counter: u16) -> Self {
+        let (seconds, nanos) = Self::rfc4122_to_unix(ticks);
+
+        #[cfg(any(feature = "v1", feature = "v6"))]
+        {
+            Timestamp {
+                seconds,
+                nanos,
+                counter: _counter,
+            }
+        }
+        #[cfg(not(any(feature = "v1", feature = "v6")))]
+        {
+            Timestamp { seconds, nanos }
+        }
+    }
+
+    /// Construct a `Timestamp` from a unix timestamp
+    ///
+    /// A unix timestamp represents the elapsed time since Jan 1 1970. Libc's
+    /// `clock_gettime` and other popular implementations traditionally
+    /// represent this duration as a `timespec`: a struct with `u64` and
+    /// `u32` fields representing the seconds, and "subsecond" or fractional
+    /// nanoseconds elapsed since the timestamp's second began,
+    /// respectively.
+    pub fn from_unix(
+        _context: impl ClockSequence<Output = u16>,
+        seconds: u64,
+        nanos: u32,
+    ) -> Self {
+        #[cfg(any(feature = "v1", feature = "v6"))]
+        {
+            let counter = _context.generate_sequence(seconds, nanos);
+            Timestamp {
+                seconds,
+                nanos,
+                counter,
+            }
+        }
+        #[cfg(not(any(feature = "v1", feature = "v6")))]
+        {
+            Timestamp { seconds, nanos }
+        }
+    }
+
+    /// Construct a `Timestamp` from the current time of day
+    /// according to Rust's SystemTime
+    #[cfg(all(feature = "std", not(any(feature = "v1", feature = "v6"))))]
+    pub fn now() -> Self {
+        let dur = std::time::SystemTime::UNIX_EPOCH
+            .elapsed()
+            .expect("Getting elapsed time since UNIX_EPOCH.  If this fails, we've somehow violated causality");
```

Comment:
I think this can happen on systems with drifting clocks, maybe more likely to be observed on Windows since its epoch is earlier than the Unix one?

I don't have a great alternative for what we should do, besides documenting that this method will panic if the system time is set earlier than the Unix epoch.

Reply:
I think that clock drift shouldn't be a problem.  The Result returned by `SystemTime::elapsed()` is an error if self's time happens **after** now.    In this case, it is checking that that the current system time is later than Jan 1st, 1970.  
I'll update the documentation and the message accordingly.

Reply:
Yeh, so less drifting clocks, and maybe more likely misconfigured Windows systems, where their epoch is some much earlier date than the Unix one 🤔